test/fake_http_server.cc

```diff
@@ -87,38 +96,68 @@ void FakeServer::TerminateAllStreams() {
-  auto stream_it = path_streams.find(request.destination);
-  if (stream_it != path_streams.end()) {
-    auto& stream = stream_it->second;
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-
-    // Create a queue for this watcher and add to the stream.
-    std::queue<std::string> my_queue;
-    stream.AddQueue(&my_queue);
-
-    // For every new string on my queue, send to the client.  The
-    // empty string indicates that we should terminate the stream.
-    while (true) {
-      std::string s = stream.GetNextResponse(&my_queue);
-      if (s.empty()) {
-        break;
+  if (request.method == "GET") {
+    auto stream_it = path_streams.find(request.destination);
+    if (stream_it != path_streams.end()) {
+      auto& stream = stream_it->second;
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+
+      // Create a queue for this watcher and add to the stream.
+      std::queue<std::string> my_queue;
+      stream.AddQueue(&my_queue);
+
+      // For every new string on my queue, send to the client.  The
+      // empty string indicates that we should terminate the stream.
+      while (true) {
+        std::string s = stream.GetNextResponse(&my_queue);
+        if (s.empty()) {
+          break;
+        }
+        connection->write(s);
       }
-      connection->write(s);
+      return;
+    }
+
+    auto it = get_handlers.find(request.destination);
+    if (it != get_handlers.end()) {
+      connection->set_status(Server::connection::ok);
+      connection->set_headers(std::map<std::string, std::string>({
+          {"Content-Type", "text/plain"},
+      }));
+      std::map<std::string, std::string> headers;
+      for (const auto& h : request.headers) {
+        headers[h.name] = h.value;
+      }
+      connection->write(it->second(request.destination, headers));
+      return;
     }
-    return;
   }
 
-  auto it = path_responses.find(request.destination);
-  if (it != path_responses.end()) {
-    connection->set_status(Server::connection::ok);
-    connection->set_headers(std::map<std::string, std::string>({
-        {"Content-Type", "text/plain"},
-    }));
-    connection->write(it->second);
-    return;
+  if (request.method == "POST") {
+    auto p = post_handlers.find(request.destination);
+    if (p != post_handlers.end()) {
+      std::map<std::string, std::string> headers;
+      std::string body;
+      connection->read([&headers, &body, &request]
+                       (Server::connection::input_range range,
+                        boost::system::error_code error,
+                        size_t size,
+                        Server::connection_ptr conn) {
+        for (const auto& h : request.headers) {
```

Comment:
Done.

Reply:
Added a comment. The callback is async, so there was a race condition where we could send the response before populating the body variable.

Reply:
You can still move the extraction of the headers out of the callback — the race condition was only on the response.
I'm also wondering why you needed the async `read` instead of just `body(request)` (or even `request.body`).

Reply:
body(request) and request.body are empty, so I have to use the async read.

Per offline discussion, added a mutex to wait for the read and moved the writes out of the callback.